connman/plugins/sailfish_ofono.c

```diff
@@ -593,6 +599,92 @@ static int modem_configure(struct modem_data *md)
 	return index;
 }
 
+static gboolean is_modem_network_configured(struct modem_data *md)
+{
+	DBG("%p", md);
+
+	switch (md->connctx->protocol) {
+	case OFONO_CONNCTX_PROTOCOL_UNKNOWN:
+	case OFONO_CONNCTX_PROTOCOL_NONE:
+		return FALSE;
+	case OFONO_CONNCTX_PROTOCOL_IP:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV4)) {
+			connman_warn("ofono: %p IPv4 no address set", md);
+			return FALSE;
+		}
+
+		break;
+	case OFONO_CONNCTX_PROTOCOL_IPV6:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV6)) {
+			connman_warn("ofono: %p IPv6 no address set", md);
+			return FALSE;
+		}
+
+		break;
+	case OFONO_CONNCTX_PROTOCOL_DUAL:
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV4)) {
+			connman_warn("ofono: %p DUAL no IPv4 address set", md);
+			return FALSE;
+		}
+
+		if (!connman_network_is_configured(md->network,
+						CONNMAN_IPCONFIG_TYPE_IPV6)) {
+			connman_warn("ofono: %p DUAL no IPv6 address set", md);
+			return FALSE;
+		}
+
+		break;
+	}
+
+	DBG("%p is configured", md);
+
+	return TRUE;
+}
+
+static gboolean delayed_connect(gpointer data) {
+	struct modem_data *md = data;
+
+	/* Keep in loop until configured and attempt limit is not reached. */
+	if (!is_modem_network_configured(md) &&
+			md->delayed_connect_attempts < DELAYED_CONNECT_LIMIT) {
+		md->delayed_connect_attempts++;
+		return TRUE;
+	}
+
+	/* Log that we're giving up, network setup wasn't completed in time. */
+	if (md->delayed_connect_attempts == DELAYED_CONNECT_LIMIT)
+		connman_error("cellular setup was not completed in time");
+
+	DBG("modem %p network %p configured, set connected", md, md->network);
+
+	connman_network_set_connected(md->network, TRUE);
+
+	g_source_remove(md->delayed_connect_id);
```

Comment:
The return FALSE (or G_SOURCE_REMOVE) makes explicit removal unnecessary.

Reply:
Yep, that was unnecessary indeed. Apparently I was too focused on the issue at hand than at the code.